Move Generic Prefixes to yaml and add prefix for Neutral

MustaphaTR opened the pull request:
Fixes #14402.

pchote commented:
Omitting the "Neutral" here was a deliberate choice.  The idea is that an is X (building/infantry/whatever) in its default state should not need any additional or redundant qualifiers.  Allied or enemy Xes have special properties that make them different to the default state, and so they are described differently.

I believe this also for consistency with the original RA and TD.

MustaphaTR commented:
> I believe this also for consistency with the original RA and TD.

Not D2K (it show Neutral prefix for Smuggler Units), maybe we should move the Prefixes to yaml.

pchote commented:
If you want to override this for D2K then you could make these stance prefixes configurable.  Replace the `GenericStancePrefix` flag with strings for `AlliedPrefix = "Allied "`, `NeutralPrefix = null`, `EnemyPrefix = "Enemy "`, and replace the `GenericStancePrefix` checks with a non-null check for the appropriate prefix.

MustaphaTR commented:
Updated. Replaced GenericStancePrefix with Ally/Neutral/EnemyPrefix. Also added a upgrade rule for it.

pchote commented on mods/cnc/rules/campaign-tooltips.yaml:
```diff
@@ -55,8 +55,9 @@
 		GenericVisibility: Enemy, Ally, Neutral
-		GenericStancePrefix: false
 		ShowOwnerRow: false
+		AllyPrefix:
```
I'm not a big fan of having to override these to empty.  IMO it would be better to keep the `GenericStancePrefix` as well as the new prefix overrides.  What do the other @OpenRA/engine-hackers think?

abcdefg30 commented:
It looks like the Fieldloader trims the white spaces, so:
> 02:21:11 (abcdefg30) does #14404 even work properly for custom names?
02:21:32 (abcdefg30) i.e. if I define "Ordos " as prefix, do we cut the trailing space away?
02:22:03 (abcdefg30) (-> the pr relies on the fieldloader not doing that)
02:22:25 (abcdefg30) if the fieldloader indeed trims the string
02:22:31 (abcdefg30) we need to hardcode a " " anyway
02:22:35 (abcdefg30) so we need that boolean again
02:22:47 (abcdefg30) to tell it "use a prefix + a whitespace" or "nothing at all"
02:23:08 (abcdefg30) (with an empty prefix we'd still have the hardcoded whitespace, if you get what I mean)

MustaphaTR committed:
Move Generic Prefixes to yaml and add prefix for Neutral

MustaphaTR commented:
Updated, this now just adds neutral and makes the prefixes settable on yaml. Check is still there, upgrade rule and changes to main mods are no longer required and removed. Also hardcoded a space into the name as @abcdefg30 said.

GraionDilach commented on OpenRA.Mods.Common/Traits/Tooltip.cs:
```diff
@@ -47,11 +56,14 @@ public string TooltipForPlayerStance(Stance stance)
 
-			if (GenericStancePrefix && stance == Stance.Ally)
-				return "Allied " + GenericName;
+			if (GenericStancePrefix && AllyPrefix != null && stance == Stance.Ally)
```
This should be string.IsNullOrEmpty I believe.

GraionDilach replied:
Applies to all three cases ofc.

GraionDilach reviewed (approved):
:+1: after that, however.

MustaphaTR commented:
Updated.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/599fb6354ac1f65a6e064ee143a236439dbc5bea
